---
title: "201-260"
author: "Mac Campbell"
date: "2/1/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message=FALSE)
```


```{r}
library(tidyverse)
library(viridis)
library(ggpubr)
library(grid)
library(gridExtra)
library(ggtree)
library(vcfR)
library(adegenet)
library(ape)
library(ggnetworx)
library(phangorn)
library(ggrepel)
```

## Create PCA
Using all chroms.    

```{sh, eval=FALSE}
srun -p med -t 12:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -P 24  -bam bamlists/test260.bamlist \
-minInd 234 -GL 1 -ref $HOME/genomes/xyrauchen/Xyrauchen_texanus.faa \
-doGLF 2 -doMajorMinor 1 -doMaf 2 -SNP_pval 1e-6 -minMapQ 10 -minQ 20 -minMaf 0.05 \
-out outputs/200/test260 > outputs/200/beagle.out 2> outputs/200/beagle.err &

python $HOME/pcangsd/pcangsd.py -beagle outputs/200/test260.beagle.gz -o outputs/200/test260 -threads 10
```

Read 260 samples and 242050 sites    
Number of sites after MAF filtering (0.05): 242043    

Generate some SNPs, 95% missing
```{sh, eval=FALSE}
srun -p med -t 24:00:00 --mem=16G --nodes=1 --ntasks=6 angsd -P 12 \
-bam bamlists/test260.bamlist \
-out /home/maccamp/wall-canyon/outputs/200/plink \
-anc $HOME/genomes/xyrauchen/Xyrauchen_texanus.faa \
-minInd 247 -minMaf 0.05  -minMapQ 10 -minQ 20 -GL 1 -doMajorMinor 1 -doMaf 1 -SNP_pval 1e-6 \
-doGeno 4 -doPost 1 -postCutoff 0.95 -doPlink 2 >outputs/200/std.out 2>outputs/200/std.err &
```

plink.tped has 19058 snps!  This is ~3000 more than with the Myxocyprinus reference.
get meta

```{r}
b4<-read_tsv("outputs/200/b4.tsv")
write_tsv(b4 %>% select(`Sample ID`), "outputs/200/samples260.tsv", col_names = FALSE)
```

```{r}
deets <- read_csv("meta/additional-meta.csv")
meta <- b4 %>% left_join(deets, by=c("Sample ID"="GVL Code"))
```

## PCA
```{r}
cov<-read_delim("outputs/200/test260.cov", col_names=FALSE, delim=" ") %>% as.matrix()
```

```{r, echo=FALSE, warning=FALSE, message=FALSE}
# Some shameless copying from Eric A.
#' @param samples character vector with the individuals IDs in the order in which
#' they were passed in the bamlist to angsd.
#' @param cov covariance matrix
covar2pcs <- function(samples, cov) {
  
  
  eig <- eigen(cov, symm = TRUE)
  PC <- as.data.frame(eig$vectors) %>%
    as_tibble() %>%
    setNames(sprintf("PC-%02d", 1:ncol(.)))
  
  samtib <- tibble(sample = samples)
  
  list(
    PCs = bind_cols(samtib, PC),
    eigevalues = eig$values
  )
}
```

```{r}
pca <- covar2pcs(meta$`Sample ID`, cov)

pca_long <- pca$PCs %>%
  tidyr::gather(., key = "PC", "val", -sample)

# then expand a grid of the possible comparisons (ordered)
expg <- expand.grid(sample = pca$PCs$sample,
                    PCx = sprintf("PC-%02d", 1:6),
                    PCy = sprintf("PC-%02d", 1:6),
                    stringsAsFactors = FALSE) %>%
  tibble::as_tibble()

# then left join the pca results onto that
pca_pairs <- dplyr::left_join(expg, pca_long, by = c("sample", "PCx" = "PC")) %>%
  dplyr::rename(val_x = val) %>%
  dplyr::left_join(pca_long, by = c("sample", "PCy" = "PC")) %>%
  dplyr::rename(val_y = val)

pp_meta <- pca_pairs %>%   # just keep the first 6 PCs around
  left_join(., meta, by = c("sample" = "Sample ID"))

# now, that has the first 6 PCs in it.  If we want to focus on the just the
# first 3, we could do 
npc <- 3
pp_meta2 <- pp_meta %>%
  filter( (PCx %in% sprintf("PC-%02d", 1:npc)) & 
            (PCy %in% sprintf("PC-%02d", 1:npc)) )

ggplot(pp_meta2, aes(x = val_x, y = val_y, fill = `Species Common Name`)) +
  geom_point(pch = 21, size = 2) +
  scale_fill_discrete(na.value = "white") + 
  facet_grid(PCy ~ PCx, scales = "free")
```

```{r}
eig <- eigen(cov, symm = TRUE)
var<-eig$values/sum(eig$values)
cumvar<-cumsum(eig$values)/sum(eig$values)

head(var)
head(cumvar)
```


```{r}
sub12<-pp_meta2 %>% filter( (PCx =="PC-01") & (PCy =="PC-02") )
centers12<- sub12 %>% group_by(`Species Common Name`) %>% mutate(meanX=mean(val_x), meanY=mean(val_y)) %>% select(`Species Common Name`, meanX, meanY) %>% unique()
pc12<-ggplot(sub12, aes(x = val_x, y = val_y, fill = `Species Common Name`)) +
  geom_point(size = 2, alpha=0.75, pch=21) +
  geom_text_repel(data=centers12, aes(x=meanX, y=meanY, label=`Species Common Name`), max.overlaps = Inf) +
  scale_fill_discrete(na.value = "white") + 
  theme_bw()+
  theme(panel.grid=element_blank())+
  xlab(paste("PC1", " ", round((100*var[1]),2), "%", sep = ""))+
  ylab(paste("PC2", " ", round((100*var[2]),2), "%", sep = ""))+
  scale_fill_viridis_d(option = "jet") +
#  scale_shape_manual("Major Group", values = c(21,22,23,24,25)) +
  theme(legend.position = "")+
  ggtitle("A") +
  theme(plot.title = element_text(face="bold", size=16))


sub13<-pp_meta2 %>% filter( (PCx =="PC-01") & (PCy =="PC-03") )
centers13<- sub13 %>% group_by(`Species Common Name`) %>% mutate(meanX=mean(val_x), meanY=mean(val_y)) %>% select(`Species Common Name`, meanX, meanY) %>% unique()
locs13<-sub13 %>% group_by(Location) %>% mutate(meanX=mean(val_x), meanY=mean(val_y)) %>% select(Location, meanX, meanY) %>% unique()

pc13<-ggplot(sub13, aes(x = val_x, y = val_y, fill=`Species Common Name`)) +
  geom_point(size = 2, alpha=0.75, pch=21) +
  geom_text_repel(data=centers13, aes(x=meanX, y=meanY, label=`Species Common Name`), max.overlaps = Inf) +
  scale_fill_discrete(na.value = "white") + 
  theme_bw()+
  theme(panel.grid=element_blank())+
  xlab(paste("PC1", " ", round((100*var[1]),2), "%", sep = ""))+
  ylab(paste("PC3", " ", round((100*var[3]),2), "%", sep = ""))+
  scale_fill_viridis_d(option = "jet") +
 # scale_shape_manual("Species Common Name", values = c(21,22,23,24,25)) +
 # guides(fill = guide_legend(override.aes=list(shape=15, color= viridis(12, option="magma")))) +
  ggtitle("B") +
  theme(plot.title = element_text(face="bold", size=16))


pcs<-ggarrange(pc12, pc13, ncol = 2, widths=c(1.1, 1.6))
pcs

ggsave("outputs/200/pcs260.pdf", width=11, height=4.5)

```

```{r}
ggplot(sub13, aes(x = val_x, y = val_y)) +
  geom_point(aes( fill=`Species Common Name`), size = 2, alpha=0.75, pch=21) +
  geom_text_repel(data=locs13, aes(x=meanX, y=meanY, label=Location), max.overlaps = Inf) +
  scale_fill_discrete(na.value = "white") + 
  theme_bw() +
  theme(panel.grid=element_blank())+
  xlab(paste("PC1", " ", round((100*var[1]),2), "%", sep = ""))+
  ylab(paste("PC3", " ", round((100*var[3]),2), "%", sep = ""))+
  scale_fill_viridis_d(option = "jet") 

ggsave("outputs/200/pcs260-locations.pdf", width=8, height=4.5)

```

## Network
Let's make a network!!!

```{sh, eval=FALSE}
plink --tped plink.tped --tfam plink.tfam  --out binary --recode --allow-extra-chr --noweb
plink --ped binary.ped --map binary.map --recode vcf --allow-extra-chr -out recode
bcftools +prune -l 0.9 -w 10000 recode.vcf  -Ov -o recode.prune.vcf
bcftools reheader --samples samples260.tsv -o recode.prune.reheadered.vcf recode.prune.vcf
```

Convert to phylip, then nexus.   
What about missing data??    

```{sh, eval=FALSE}
source activate py2; ~/github/mccloud-rrt/vcf2phylip.py -i recode.prune.reheadered.vcf; conda deactivate;
seqConverter.pl -drecode.prune.reheadered.min4.phy -on

#Corrected for ascertainment bias
~/github/mccloud-rrt/103-remove-invariant.py -p recode.prune.reheadered.min4.phy -o recode.prune.reheadered.min4.asc.phy
seqConverter.pl -drecode.prune.reheadered.min4.asc.phy -on

#Now remove missing??
 ~/github/relict-dace/201.1-removeMissing.pl recode.prune.reheadered.min4.asc.phy  > filtered.phy
seqConverter.pl -dfiltered.phy -on

```
    

```{r}
data<-read.nexus.data("outputs/200/recode.prune.reheadered.min4.nex")
write.nexus.data(data, file="outputs/200/test260.nex")

data<-read.nexus.data("outputs/200/recode.prune.reheadered.min4.asc.nex")
write.nexus.data(data, file="outputs/200/test260.asc.nex")

data<-read.nexus.data("outputs/200/filtered.nex")
write.nexus.data(data, file="outputs/200/test260.filtered.nex")
```

```{r}
network<-read.nexus.networx(file = "outputs/200/test260-network.nex", splits=FALSE)
pdf(file = "outputs/200/test260-network.pdf", width=20, height=20)
plot(network, edge.width = 0.5)
dev.off()
```

```{r}
network<-read.nexus.networx(file = "outputs/200/test260-network.asc.nex", splits=FALSE)
pdf(file = "outputs/200/test260-network.asc.pdf", width=20, height=20)
plot(network, edge.width = 0.5)
dev.off()
```

```{r}
network<-read.nexus.networx(file = "outputs/200/test260-network.filtered.nex", splits=FALSE)
pdf(file = "outputs/200/test260-network.filtered.pdf", width=20, height=20)
plot(network, edge.width = 0.5)
dev.off()
```

Make it more readable?

Consensus network

```{r}
dna<-read.phyDat("outputs/200/filtered.phy")
dat<-as.phyDat(dna)
set.seed(1)
bs <- bootstrap.phyDat(dat, FUN = function(x)nj(dist.hamming(x)), 
    bs=100)
tree <- nj(dist.hamming(dat))
#par("mar" = rep(1, 4))
#tree<-root(tree, c("PCT_PPP1_01\t","LCT_GACR_10\t"))
#tree <- plotBS(tree, bs, "phylogram")

#By default prob=0.3
cnet <- consensusNet(bs, .25)
cnet$tip.label<-gsub("\t","",cnet$tip.label)

#plot(cnet, "2D", show.edge.label=TRUE)

edge.col <- createLabel(cnet, tree, "black", nomatch="red")

plot(cnet, show.edge.label = T, "2D", edge.color = edge.col,
                  col.edge.label = "blue", cex=.75)
```


```{r}
pdf(file = "outputs/200/test260-consensus-network.filtered.pdf", width=20, height=20)
plot(cnet, show.edge.label = T, "2D", edge.color = edge.col,
                col.edge.label = "blue", cex=.75)
dev.off()
```