---
title: "204-targeted-phylogeny"
author: "Mac Campbell"
date: "4/14/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message=FALSE)
```


```{r}
library(tidyverse)
library(ggtree)
```

## Phylogeny

Generating a data set for concated and species tree analyses. Will reduce sampling of non-wow clade fish.

These our are downsampled and complete read counts.

```{r}
b3b<-read_tsv("meta/b3b.tsv")
```

### Mountains
Get 2-4 or each location of Pantosteus

```{r}
mountains<-read_tsv("outputs/300/mymountains.tsv")
tahoe<-read_tsv("outputs/300/mytahoe.tsv")
```

```{r}
mpc<-b3b %>% select(`Sample ID`, `Species Common Name`, Location, Counts, NewPath) %>% filter(`Sample ID` %in% mountains$sample) %>% group_by(Location) %>% arrange(-Counts) %>%
  filter(Location=="Poore Creek, CA") %>% top_n(n=2, wt=Counts)

mtr<-b3b %>% select(`Sample ID`, `Species Common Name`, Location, Counts, NewPath) %>% filter(`Sample ID` %in% mountains$sample) %>% group_by(Location) %>% arrange(-Counts) %>%
  filter(Location=="Truckee River, NV") %>% top_n(n=2, wt=Counts)
```

```{r}
og<-bind_rows(mpc,mtr)
```


### Tahoes
```{r}
tpc<-b3b %>% select(`Sample ID`, `Species Common Name`, Location, Counts, NewPath) %>% filter(`Sample ID` %in% tahoe$sample) %>% group_by(Location) %>% arrange(-Counts) %>%
  filter(Location=="Poore Creek, CA") %>% top_n(n=2, wt=Counts)

ttr<-b3b %>% select(`Sample ID`, `Species Common Name`, Location, Counts, NewPath) %>% filter(`Sample ID` %in% tahoe$sample) %>% group_by(Location) %>% arrange(-Counts) %>%
  filter(Location=="Truckee River, NV") %>% top_n(n=2, wt=Counts)

tcr<-b3b %>% select(`Sample ID`, `Species Common Name`, Location, Counts, NewPath) %>% filter(`Sample ID` %in% tahoe$sample) %>% group_by(Location) %>% arrange(-Counts) %>%
  filter(Location=="East Fork Carson River, NV") %>% top_n(n=2, wt=Counts)

```

```{r}
tahoes<-bind_rows(tpc,ttr,tcr)
```

These are in the 259-svdq.nex tree.

### Sacs

```{r}
additional<-read_csv("meta/additional-meta.csv")

lac<- b3b %>% left_join(additional, by=c("Sample ID"="GVL Code")) %>% select(`Sample ID`, `Species Common Name`, Location, Counts, NewPath, Subspecies) %>% filter(Subspecies=="lacuanserinus") %>%
  arrange(-Counts)  %>% top_n(n=2, wt=Counts) %>% select(-Subspecies)

# lac from Ash Creek has generally good coverage, Goose lake dent has 156708 reads (1208-008)

occ<-b3b %>% left_join(additional, by=c("Sample ID"="GVL Code")) %>% select(`Sample ID`, `Species Common Name`, Location, Counts, NewPath, Subspecies) %>% filter(Subspecies=="occidentalis") %>%
  arrange(-Counts)  %>% top_n(n=2, wt=Counts) %>% select(-Subspecies)

# occidentalis also has low reads.... 
```

```{r}
sacs<-bind_rows(lac, occ)
```

### Modocs
```{r}
modocs<-b3b %>% left_join(additional, by=c("Sample ID"="GVL Code")) %>% select(`Sample ID`, `Species Common Name`, Location, Counts, NewPath, Coordinates) %>%
  filter(`Species Common Name`=="Modoc Sucker") %>% 
  arrange(-Counts) %>% top_n(4, Counts) %>% select(-Coordinates)

#Two turner creeks and two ash creeks
```


### Klams

Get a couple of each species
```{r}
larges<-b3b %>% select(`Sample ID`, `Species Common Name`, Location, Counts, NewPath) %>% filter(`Species Common Name`=="Klamath Largescale Sucker") %>%
  arrange(-Counts)  %>% top_n(n=2, wt=Counts)

smalls<-b3b %>% select(`Sample ID`, `Species Common Name`, Location, Counts, NewPath) %>% filter(`Species Common Name`=="Klamath Smallscale Sucker") %>%
  arrange(-Counts)  %>% top_n(n=2, wt=Counts)

losts<-b3b %>% select(`Sample ID`, `Species Common Name`, Location, Counts, NewPath) %>% filter(`Species Common Name`=="Lost River Sucker") %>%
  arrange(-Counts)  %>% top_n(n=2, wt=Counts)

shorts<-b3b %>% select(`Sample ID`, `Species Common Name`, Location, Counts, NewPath) %>% filter(`Species Common Name`=="Shortnose Sucker") %>%
  arrange(-Counts)  %>% top_n(n=2, wt=Counts)
```

```{r}
klams<-bind_rows(larges,smalls,losts,shorts)
```


### Wow Clade
```{r}
wowsers<-b3b %>% select(`Sample ID`, `Species Common Name`, Location, Counts, NewPath) %>% filter(`Species Common Name` %in% c("Wall Canyon Sucker","Warner Sucker","Owens Sucker")) %>%
  arrange(-Counts) 
  
ggplot(wowsers) +
  geom_histogram(aes(x = Counts, fill=`Species Common Name`)) +
  theme_bw()
```

```{r}
wowsers %>% filter(Counts>5e5) %>% group_by(`Species Common Name`,Location) %>% summarize(Count=n())
```

```{r}
keepers<-wowsers %>% filter(Counts>4e5)
```


## Combine

```{r}
myfish<-bind_rows(og,tahoes,sacs,modocs,klams,keepers)

myfish %>% group_by(`Species Common Name`) %>% summarize(Count=n())
```


```{r}
nrow(myfish)
```


```{r}
write_tsv(select(myfish, NewPath), "bamlists/phylo119.bamlist")
```
```{sh, eval=FALSE}
srun -p bigmemh -t 18:00:00 --mem=16G --nodes=1 --ntasks=1 angsd -P 4 \
-bam bamlists/phylo119.bamlist \
-out /home/maccamp/wall-canyon/outputs/204/plink \
-anc $HOME/genomes/xyrauchen/Xyrauchen_texanus.faa \
-minInd 113 -minMaf 0.05  -minMapQ 10 -minQ 20 -GL 1 -doMajorMinor 1 -doMaf 1 -SNP_pval 1e-6 \
-doGeno 4 -doPost 1 -postCutoff 0.95 -doPlink 2 >outputs/204/std.out 2>outputs/204/std.err &
```
