---
title: "800-klamath-sacremento"
author: "Mac Campbell"
date: "4/11/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message=FALSE)
```

```{r}
library(tidyverse)
library(ggrepel)
library(ggpubr)
```

## Samples sequenced

 `Common Name`     Subspecies    Coordinates                Count      
  <chr>             <chr>         <chr>                      <int>    
1 Modoc Sucker      NA            Ash Creek: Johnson            10     
2 Modoc Sucker      NA            Turner Creek: Garden Gulch    10     
3 Sacremento Sucker lacuanserinus Cottonwood                     9     
4 Sacremento Sucker lacuanserinus Goose Lake: Dent               5    
5 Sacremento Sucker occidentalis  Ash Creek: Ash Valley         13    

q's: Are all Sacramento Suckes most-closely related to each other?  Thomas Creek in Goose Lake Basin is the third location of Modoc that exists and we don't have it. But, we can see if Cottonwood/Dent are more closely related to Ash (but expecting hybrids)

I expect the S. lacuanserinus + Modoc Sucker to be most closely related to each other. 

```{r}
b3b<-read_tsv("meta/b3b.tsv") 
mm<-read_csv("meta/additional-meta.csv") %>% select(`GVL Code`,Coordinates) %>% rename(`Sample ID` = `GVL Code`)
sacs<- b3b %>% filter(`Species Common Name` %in% c("Modoc Sucker","Sacremento Sucker")) %>% left_join(mm)
```

```{r}
sacs %>% filter(Counts > 5e4) %>% select(`Sample ID`, `Species Common Name`, Counts, Coordinates) 
```
```{r}
sacs %>% filter(Counts > 5e4) %>% select(`Sample ID`, `Species Common Name`, Counts, Coordinates) %>%
  group_by(`Species Common Name`, Coordinates) %>% summarize(Count=n())
``` 
```{r}
ggplot(sacs %>% filter(Counts > 5e4)) +
  geom_histogram(aes(x=Counts, fill=`Species Common Name`)) +
  xlim(0, 1.5e6)
```


## doIBs

```{r}
r1<-sacs %>% filter(Counts > 5e4)
write_tsv(r1 %>% select(NewPath), col_names = FALSE, file="bamlists/sacs33.bamlist")
```


```{sh, eval=FALSE}
srun -p med -t 2:00:00 --mem=32G --nodes=1 $HOME/angsd/angsd -P 36  \
-bam bamlists/sacs33.bamlist  -r 'lcl|RBS_Chr1' \
-ref /$HOME/genomes/xyrauchen/Xyrauchen_texanus.faa \
-minInd 17 -minMapQ 10 -minQ 20 -GL 2 -doGLF 2 \
-doMajorMinor 1 -doMaf 1 -SNP_pval 1e-3 \
-doIBS 1 -doCounts 1 -doCov 1 -makeMatrix 1 -minMaf 0.05 \
-out outputs/800/33-chrom01-ibs > outputs/800/33-chrom01-ibs.stdout 2> outputs/800/33-chrom01-ibs.stderr &

srun -p med -t 2:00:00 --mem=32G --nodes=1 $HOME/angsd/angsd -P 36  \
-bam bamlists/sacs33.bamlist \
-ref /$HOME/genomes/xyrauchen/Xyrauchen_texanus.faa \
-minInd 17 -minMapQ 10 -minQ 20 -GL 2 -doGLF 2 \
-doMajorMinor 1 -doMaf 1 -SNP_pval 1e-3 \
-doIBS 1 -doCounts 1 -doCov 1 -makeMatrix 1 -minMaf 0.05 \
-out outputs/800/33-ibs > outputs/800/33-ibs.stdout 2> outputs/800/33-ibs.stderr &

srun -p med -t 2:00:00 --mem=32G --nodes=1 $HOME/angsd/angsd -P 36  \
-bam bamlists/sacs33.bamlist \
-ref /$HOME/genomes/xyrauchen/Xyrauchen_texanus.faa \
-minInd 30 -minMapQ 10 -minQ 20 -GL 2 -doGLF 2 \
-doMajorMinor 1 -doMaf 1 -SNP_pval 1e-3 \
-doIBS 1 -doCounts 1 -doCov 1 -makeMatrix 1 -minMaf 0.05 \
-out outputs/800/33-ibs-90 > outputs/800/33-ibs.stdout 2> outputs/800/33-ibs.stderr &
```

6,601 snps on chrom01
191666 total 50% missing
28815 at 90% missing 
```{r}
m <- as.matrix(read.table("outputs/800/33-ibs.covMat"))
meta<-r1
eig <- eigen(m)
var<-eig$values/sum(eig$values)
cumvar<-cumsum(eig$values)/sum(eig$values)

head(var)
head(cumvar)
covs<-eig$vectors[,1:3] %>% as_tibble() %>% bind_cols(meta)

text12<-covs %>% select(`Sample ID`, `Species Common Name`, Coordinates, V1, V2) %>%
  group_by(`Species Common Name`, Coordinates,) %>% summarize(Count=n(), x=mean(V1), y=mean(V2))

covs12<-ggplot(covs) +
  geom_point(aes(x=V1, y=V2, fill=`Species Common Name`), pch=21, alpha=0.75) +
  geom_text_repel(data=text12, aes(x=x, y=y, label=Coordinates), max.overlaps = Inf) +
  xlab(paste("PC1", " ", round((100*var[1]),2), "%", sep = "")) +
  ylab(paste("PC2", " ", round((100*var[2]),2), "%", sep = "")) +
  theme_bw() +
  theme(legend.position="")


text13<-covs %>% select(`Sample ID`, `Species Common Name`,Coordinates, V1, V3) %>%
 group_by(`Species Common Name`, Coordinates) %>% summarize(Count=n(), x=mean(V1), y=mean(V3))

covs13<-ggplot(covs) +
  geom_point(aes(x=V1, y=V3, fill=`Species Common Name`), pch=21, alpha=0.75) +
  geom_text_repel(data=text13, aes(x=x, y=y, label=Coordinates), max.overlaps = Inf) +
  xlab(paste("PC1", " ", round((100*var[1]),2), "%", sep = "")) +
  ylab(paste("PC3", " ", round((100*var[3]),2), "%", sep = "")) +
  theme_bw() 

ggarrange(covs12, covs13, widths = c(1,1.25))

ggsave("outputs/800/sac-33-pcs.jpeg", width=10, height=5)
```

## Outgroup + Call SNPs?

```{r}
myfish<-read.csv("meta/phylo119.csv")  %>% filter(Species.Common.Name %in% c("Klamath Largescale Sucker","Klamath Smallscale Sucker","Lost River Sucker","Shortnose Sucker")) %>%
  mutate(Coordinates="Klamath Basin") %>% select(Sample.ID, Species.Common.Name,Counts,Location,Coordinates, NewPath ) %>%
  rename(`Sample ID`=Sample.ID, `Species Common Name`=Species.Common.Name)

r2<-select(r1,`Sample ID`, `Species Common Name`, Counts, Location, Coordinates, NewPath) %>% 
  bind_rows(myfish) 

write_tsv(r2 %>% select(NewPath), col_names = FALSE, file="bamlists/sacs41.bamlist")
write_tsv(r2 %>% select(`Sample ID`), col_names = FALSE, file="bamlists/sacs41.names")

```

```{sh, eval=FALSE}
srun -p med -t 2:00:00 --mem=32G --nodes=1 $HOME/angsd/angsd -P 36  \
-bam bamlists/sacs41.bamlist \
-ref /$HOME/genomes/xyrauchen/Xyrauchen_texanus.faa \
-minInd 31 -minMapQ 10 -minQ 20 -GL 2 -doGLF 2 \
-doMajorMinor 1 -doMaf 1 -SNP_pval 1e-3 \
-doIBS 1 -doCounts 1 -doCov 1 -makeMatrix 1 -minMaf 0.05 \
-out outputs/800/41-ibs-75 > outputs/800/41-ibs.stdout 2> outputs/800/41-ibs.stderr &

srun -p med -t 2:00:00 --mem=32G --nodes=1 $HOME/angsd/angsd -P 36  \
-bam bamlists/sacs41.bamlist  -r 'lcl|RBS_Chr1' \
-ref /$HOME/genomes/xyrauchen/Xyrauchen_texanus.faa \
-minInd 31 -minMapQ 10 -minQ 20 -GL 2 -doGLF 2 \
-doMajorMinor 1 -doMaf 1 -SNP_pval 1e-3 \
-doIBS 1 -doCounts 1 -doCov 1 -makeMatrix 1 -minMaf 0.05 \
-out outputs/800/41-ibs-chrom01-75 > outputs/800/41-ibs-chrom01.stdout 2> outputs/800/41-ibs-chrom01.stderr &
```

115749 genome-wide SNPS

```{r}
m <- as.matrix(read.table("outputs/800/41-ibs-75.covMat"))
meta<-r2
eig <- eigen(m)
var<-eig$values/sum(eig$values)
cumvar<-cumsum(eig$values)/sum(eig$values)

head(var)
head(cumvar)
covs<-eig$vectors[,1:3] %>% as_tibble() %>% bind_cols(meta)

text12<-covs %>% select(`Sample ID`, `Species Common Name`, Coordinates, V1, V2) %>%
  group_by(`Species Common Name`, Coordinates,) %>% summarize(Count=n(), x=mean(V1), y=mean(V2))

covs12<-ggplot(covs) +
  geom_point(aes(x=V1, y=V2, fill=`Species Common Name`), pch=21, alpha=0.75) +
  geom_text_repel(data=text12, aes(x=x, y=y, label=Coordinates), max.overlaps = Inf) +
  xlab(paste("PC1", " ", round((100*var[1]),2), "%", sep = "")) +
  ylab(paste("PC2", " ", round((100*var[2]),2), "%", sep = "")) +
  theme_bw() +
  theme(legend.position="")


text13<-covs %>% select(`Sample ID`, `Species Common Name`,Coordinates, V1, V3) %>%
 group_by(`Species Common Name`, Coordinates) %>% summarize(Count=n(), x=mean(V1), y=mean(V3))

covs13<-ggplot(covs) +
  geom_point(aes(x=V1, y=V3, fill=`Species Common Name`), pch=21, alpha=0.75) +
  geom_text_repel(data=text13, aes(x=x, y=y, label=Coordinates), max.overlaps = Inf) +
  xlab(paste("PC1", " ", round((100*var[1]),2), "%", sep = "")) +
  ylab(paste("PC3", " ", round((100*var[3]),2), "%", sep = "")) +
  theme_bw() 

ggarrange(covs12, covs13, widths = c(1,1.5))

ggsave("outputs/800/sac-41-klams-pcs.jpeg", width=10, height=5)
```
## Add some of each wow

```{r}
my2<-read_csv("meta/phylo119.csv") %>% filter(`Species Common Name` %in% c("Owens Sucker")) %>%
  mutate(Coordinates="Owens Valley") %>% select(`Sample ID`, `Species Common Name`,Counts,Location,Coordinates, NewPath ) %>% group_by(`Species Common Name`) %>%
  top_n(2,Counts) 

r3<-bind_rows(r2,my2)
r3

write_tsv(r3 %>% select(NewPath), col_names = FALSE, file="bamlists/sacs43.bamlist")
write_tsv(r3 %>% select(`Sample ID`), col_names = FALSE, file="bamlists/sacs43.names")

```

```{sh, eval=FALSE}
srun -p med -t 2:00:00 --mem=32G --nodes=1 $HOME/angsd/angsd -P 36  \
-bam bamlists/sacs43.bamlist \
-ref /$HOME/genomes/xyrauchen/Xyrauchen_texanus.faa \
-minInd 32 -minMapQ 10 -minQ 20 -GL 2 -doGLF 2 \
-doMajorMinor 1 -doMaf 1 -SNP_pval 1e-3 \
-doIBS 1 -doCounts 1 -doCov 1 -makeMatrix 1 -minMaf 0.05 \
-out outputs/800/43-ibs-75 > outputs/800/43-ibs.stdout 2> outputs/800/43-ibs.stderr &
``` 


138,113 sites   


```{r}
m <- as.matrix(read.table("outputs/800/43-ibs-75.covMat"))
meta<-r3
eig <- eigen(m)
var<-eig$values/sum(eig$values)
cumvar<-cumsum(eig$values)/sum(eig$values)

head(var)
head(cumvar)
covs<-eig$vectors[,1:3] %>% as_tibble() %>% bind_cols(meta)

text12<-covs %>% select(`Sample ID`, `Species Common Name`, Coordinates, V1, V2) %>%
  group_by(`Species Common Name`, Coordinates,) %>% summarize(Count=n(), x=mean(V1), y=mean(V2))

covs12<-ggplot(covs) +
  geom_point(aes(x=V1, y=V2, fill=`Species Common Name`), pch=21, alpha=0.75) +
  geom_text_repel(data=text12, aes(x=x, y=y, label=Coordinates), max.overlaps = Inf) +
  xlab(paste("PC1", " ", round((100*var[1]),2), "%", sep = "")) +
  ylab(paste("PC2", " ", round((100*var[2]),2), "%", sep = "")) +
  theme_bw() +
  theme(legend.position="")


text13<-covs %>% select(`Sample ID`, `Species Common Name`,Coordinates, V1, V3) %>%
 group_by(`Species Common Name`, Coordinates) %>% summarize(Count=n(), x=mean(V1), y=mean(V3))

covs13<-ggplot(covs) +
  geom_point(aes(x=V1, y=V3, fill=`Species Common Name`), pch=21, alpha=0.75) +
  geom_text_repel(data=text13, aes(x=x, y=y, label=Coordinates), max.overlaps = Inf) +
  xlab(paste("PC1", " ", round((100*var[1]),2), "%", sep = "")) +
  ylab(paste("PC3", " ", round((100*var[3]),2), "%", sep = "")) +
  theme_bw() 

ggarrange(covs12, covs13, widths = c(1,1.5))

ggsave("outputs/800/sac-43-klams-pcs.jpeg", width=10, height=5)
```

Make some calls:

```{sh, eval=FALSE}
srun -p high -t 4:00:00 --mem=16G --nodes=1 angsd -P 24 \
-bam bamlists/sacs43.bamlist \
-out /home/maccamp/wall-canyon/outputs/800/plink \
-anc $HOME/genomes/xyrauchen/Xyrauchen_texanus.faa \
-minInd 38 -minMaf 0.05  -minMapQ 10 -minQ 20 -GL 1 -doMajorMinor 1 -doMaf 1 -SNP_pval 1e-3 \
-doGeno 4 -doPost 1 -postCutoff 0.95 -doPlink 2 >outputs/800/std.out 2>outputs/800/std.err &
```

1693 sites with 6 wow clade members     
7909 sites with 2 wow clade members


```{sh, eval=FALSE}
plink --tped plink.tped --tfam plink.tfam  --out binary --recode --allow-extra-chr --noweb
plink --ped binary.ped --map binary.map --recode vcf --allow-extra-chr -out recode
bcftools +prune -l 0.9 -w 10000 recode.vcf  -Ov -o recode.prune.vcf
bcftools reheader --samples ../../bamlists/sacs43.names -o recode.prune.reheadered.vcf recode.prune.vcf

```

```{sh, eval=FALSE}
source activate py2; ~/github/mccloud-rrt/vcf2phylip.py -i recode.prune.reheadered.vcf; conda deactivate;

 ~/github/relict-dace/201.1-removeMissing.pl recode.prune.reheadered.min4.phy  > filtered-min4.phy
seqConverter.pl -drecode.prune.reheadered.min4.phy -on

#Corrected for ascertainment bias
conda activate py3; ~/github/mccloud-rrt/103-remove-invariant.py -p recode.prune.reheadered.min4.phy -o recode.prune.reheadered.min4.asc.phy; conda deactivate

seqConverter.pl -drecode.prune.reheadered.min4.asc.phy -on

#Now remove missing?? 
 ~/github/relict-dace/201.1-removeMissing.pl recode.prune.reheadered.min4.asc.phy  > filtered.phy
seqConverter.pl -dfiltered.phy -on
```

```{sh, eval=FALSE}
iqtree -s filtered-min4.phy -st DNA -m GTR+F+ASC -bb 1000 -alrt 1000 -redo
iqtree -s filtered-min4.phy.varsites.phy -st DNA -m GTR+F+ASC -bb 1000 -alrt 1000 -redo -T AUTO
```

Getting rid of low counts. 

```{r}
r4<-r3 %>% group_by(`Species Common Name`, Coordinates) %>% top_n(4,Counts) %>%
  filter(Counts > 1.5e5)

write_tsv(r4 %>% select(NewPath), col_names = FALSE, file="bamlists/sacs24.bamlist")
write_tsv(r4 %>% select(`Sample ID`), col_names = FALSE, file="bamlists/sacs24.bamlist")
```

## NGSDist version

```{sh, eval=FALSE}
NSITES=`zcat outputs/800/43-ibs-75.mafs.gz| tail -n+2 | wc -l`
echo $NSITES

#Pop label example (I already have lables in meta/phylo-sample-names.txt)
#Rscript -e 'cat(paste(rep(c("LWK","TSI","PEL"),each=10), rep(1:10, 3), sep="_"), sep="\n", file="outputs/104/pops.label")'
#cat outputs/104/pops.label

module load ngsTools/07-2019

srun -p high --nodes=1 -t 01:00:00 ngsDist -verbose 1 -geno outputs/800/43-ibs-75.beagle.gz -probs -n_ind 43 -n_sites $NSITES -labels bamlists/sacs43.names -o outputs/800/phylo.dist -n_threads 4 
less -S outputs/800/phylo.dist

module load fastme/2.1.5     

srun -p high --nodes=1 -t 01:00:00 fastme -D 1 -i outputs/800/phylo.dist -o outputs/800/phylo.tree -m b -n b
```

```{sh,eval=FALSE}
NSITES=`zcat outputs/800/41-ibs-75.mafs.gz| tail -n+2 | wc -l`
echo $NSITES

#Pop label example (I already have lables in meta/phylo-sample-names.txt)
#Rscript -e 'cat(paste(rep(c("LWK","TSI","PEL"),each=10), rep(1:10, 3), sep="_"), sep="\n", file="outputs/104/pops.label")'
#cat outputs/104/pops.label

module load ngsTools/07-2019

srun -p high --nodes=1 -t 01:00:00 ngsDist -verbose 1 -geno outputs/800/41-ibs-75.beagle.gz -probs -n_ind 41 -n_sites $NSITES -labels bamlists/sacs41.names -o outputs/800/phylo.dist -n_threads 4 
less -S outputs/800/phylo.dist

module load fastme/2.1.5     

srun -p high --nodes=1 -t 01:00:00 fastme -D 1 -i outputs/800/phylo.dist -o outputs/800/phylo.tree -m b -n b
```


Blargh

Get higher counts nad work with that.